### DISCUSSION TITLE
Proprietary Codecs in Qutebrowser with QT6: some work some don't

### DISCUSSION THREAD

**Discussion by vnckppl** [qutebrowser/qutebrowser#8060] [Q&A]
# Background
I recently switched from qutebrowser with qt5 installed via apt-get on Ubuntu 22.04 to qutetbrowser with qt6 installed from Github in a virtual enviornment.

## Install script
```shell
# * Install qt6
sudo apt-get update && sudo apt-get install qt6-base-dev

# * Clone the qutebrowser github
cd /home/vincent/Software || exit
git clone https://github.com/qutebrowser/qutebrowser.git
cd qutebrowser || exit

# * Install python virtual env
pversion=$(python3 --version | awk '{ print $2 }' | cut -d. -f1-2)
sudo apt-get install "python${pversion}-venv"

# * Install pyqt6
python3 -m pip install PyQt6 PyQt6-WebEngine

# * Set pyqt6 in the environment
export QUTE_QT_WRAPPER=PyQt6

# * Install virtual environment
python3 scripts/mkvenv.py --pyqt-type link
``` 

## Qutebrowser version
```
qutebrowser v3.1.0
Git commit: b490f288a on main (2024-01-01 19:27:03 +1300)
Backend: QtWebEngine 6.6.1, based on Chromium 112.0.5615.213 (from api)
Qt: 6.6.1

CPython: 3.10.12
PyQt: 6.6.1

Qt wrapper info:
  PyQt6: success
  PyQt5: not imported
  -> selected: PyQt6 (via autoselect)

colorama: 0.4.6
jinja2: 3.1.2
pygments: 2.17.2
yaml: 6.0.1
adblock: 0.6.0
objc: no
PyQt6.QtWebEngineCore: 6.6.0
PyQt6.sip: 6.7.12
pdf.js: no
sqlite: 3.44.0
QtNetwork SSL: OpenSSL 3.0.2 15 Mar 2022

Style: QFusionStyle
Platform plugin: xcb
OpenGL: Intel, 4.6 (Compatibility Profile) Mesa 23.1.3
Platform: Linux-6.5.0-10013-tuxedo-x86_64-with-glibc2.35, 64bit
Linux distribution: Ubuntu 22.04.3 LTS (ubuntu)
Frozen: False
Imported from /home/vincent/Software/qutebrowser/qutebrowser
Using Python from /home/vincent/Software/qutebrowser/.venv/bin/python
Qt library executable path: /home/vincent/Software/qutebrowser/.venv/lib/python3.10/site-packages/PyQt6/Qt6/libexec, data path: /home/vincent/Software/qutebrowser/.venv/lib/python3.10/site-packages/PyQt6/Qt6

Paths:
cache: /home/vincent/.cache/qutebrowser
config: /home/vincent/.config/qutebrowser
data: /home/vincent/.local/share/qutebrowser
runtime: /run/user/1000/qutebrowser
``` 

# Issue
Not all media is playing. For example, Youtube and Soundcloud play audio and video without issues, but Spotify or Invidious do no work at all. 

# Question
Are there any additional packages I need to install to enable playback of more formats/sites in qutebrowser with qt6?

**Comment by The-Compiler** [comment]
Qt binaries come without proprietary codec support. This is a build-time option, so there's no way to install them after the fact.

**Reply by vnckppl** [reply]
I understood that with `--pyqt-type link`, proprietary codecs would work. Also, as I wrote, some of them are working (youtube and soundcloud). Is there anything I can do before building the qutebroswer venv to have the other codecs also work?

**Reply by The-Compiler** [reply]
Those websites aren't using proprietary codecs then. And no, other than using your distribution's Qt or building your own, there's no way around this.

**Comment by toofar** [comment]
> python3 -m pip install PyQt6 PyQt6-WebEngine
> Qt library executable path: /home/vincent/Software/qutebrowser/.venv/lib/python3.10/site-packages/PyQt6/Qt6/libexec

PyQt is linked against a Qt library. If you want to use your distro Qt I think you should be using the PyQt from your system (which the --link argument should handle), so try not installing PyQt6 and PyQt6-WebEngine to the venv. Assuming 22.04 has Qt6 packaged.

**Reply by dmatos2012** [reply]
Has anyone managed to build it (qtwebengine 6.4) on `Ubuntu 20.04` ? I really want at least youtube live or some other media to play, but so far it looks almost impossible to build from source. I find way too many different ways to do so, either in `qt` website or "tutorials". And for sure, I do not want to build all of `qt` since it will be so massive.

  Any time I watch anything live, have to switch browsers. 

Thank you!

**Reply by vnckppl** [reply]
That first python line is not installing PyQt6 into the virtual environment, but system wide. 

This is also apparent when I run the mkvenv.py script, as it indicates it is symlinking this local python library to inside the virtual environement:

```
==================== Linking system-wide PyQt ====================
/home/vincent/.local/lib/python3.10/site-packages/PyQt6 -> /home/vincent/Software/qutebrowser/.venv/lib/python3.10/site-packages/PyQt6
```

I also have qt6-webengine-dev installed via apt-get:
`sudo apt-get update && sudo apt-get install qt6-base-dev qt6-webengine-dev `

So, the updated script would be:
``` bash
# * Install qt6
sudo apt-get update && sudo apt-get install qt6-base-dev qt6-webengine-dev

# * Clone the qutebrowser github
cd /home/vincent/Software || exit
git clone https://github.com/qutebrowser/qutebrowser.git
cd qutebrowser || exit

# * Install pyqt6 (locally)
python3 -m pip install PyQt6 PyQt6-WebEngine

# * Install virtual environment
python3 scripts/mkvenv.py --pyqt-type link
```

However, that did not resolve the issue.

Suggestions are welcome.

**Reply by The-Compiler** [reply]
Don't install PyQt via pip, there's no point in doing that and linking to it vs. just installing it in the virtualenv. Install Qt and PyQt via apt instead.
